MonkeyWrench.Web.WebService/Scheduler/Scheduler.cs

```diff
@@ -400,11 +395,43 @@ private static void AddWork (DB db, List<DBHost> hosts, List<DBLane> lanes, List
 								}
 							}
 
-							sql.AppendFormat ("UPDATE RevisionWork SET state = {0} WHERE id = {1} AND state = 10;", (int) (has_dependencies ? DBState.DependencyNotFulfilled : DBState.NotDone), revisionwork.id);
+							sql.AppendFormat ("UPDATE RevisionWork SET state = 11 WHERE id = {0} AND state = 10;", revisionwork.id);
+							if (!editedHostLanes.Contains (hostlane)) editedHostLanes.Add (hostlane);
+						}
+					}
+				}
 
+				foreach (var editedHostLane in editedHostLanes) {
+					int mostRecent = -1;
+					int mostRecentState = -1;
+					using (IDbCommand cmd = db.CreateCommand ()) {
+						cmd.CommandText = @"
+SELECT RevisionWork.id, RevisionWork.state
+FROM RevisionWork
+INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
+WHERE
+     RevisionWork.host_id = @host_id
+ AND RevisionWork.lane_id = @lane_id
+ORDER BY Revision.Date DESC
+LIMIT 1
+";
+						DB.CreateParameter (cmd, "host_id", editedHostLane.host_id);

```

Comment:
You're right, my query is completely wrong.

But I think I have it now:

```sql
SELECT RevisionWork.id, Lane.lane, Revision.Revision, Host.host
FROM RevisionWork
INNER JOIN Lane ON Lane.id = RevisionWork.lane_id
INNER JOIN Revision ON Revision.id = RevisionWork.revision_id
INNER JOIN Host ON Host.id = RevisionWork.host_id
INNER JOIN (
	SELECT a.lane_id, a.date, a.id 
	FROM Revision a
	INNER JOIN (
		SELECT lane_id, MAX (date) date 
		FROM Revision 
		GROUP BY lane_id
	) b ON a.lane_id = b.lane_id AND a.date = b.date
	INNER JOIN Lane ON Lane.id = a.lane_id
	WHERE Lane.enabled = true
) LATEST ON RevisionWork.lane_id = latest.lane_id AND RevisionWork.revision_id = latest.id;

    Execution time: 536.218 ms
```

which seems to give the right results: https://gist.github.com/rolfbjarne/7646f63c9ecb1d9e8b9a02b73638ddd8

Filtering to a single lane_id (instead of all enabled lanes), made the query take 1.5ms instead of 500ms, so it might become even faster if filtering to all applicable lanes instead of all enabled lanes.